File: frontend/src/lib/agentCards.ts
import type { AgentToolRun, KnowledgeEntityType, KnowledgeSource } from "@/lib/types";

const KINDS: Record<string, KnowledgeEntityType> = { task: "task", document: "document", risk: "risk", milestone: "milestone", node: "wbs_node", stage: "stage", project: "project", sticker: "sticker", member: "member", attachment: "attachment", comment: "comment" };
const text = (value: unknown): string | null => typeof value === "string" && value.trim() ? value : null;
const record = (value: unknown): Record<string, unknown> => value && typeof value === "object" && !Array.isArray(value) ? value as Record<string, unknown> : {};

/** Старые результаты действий тоже отображаются карточками, без миграции истории. */
export function actionSources(action: AgentToolRun): KnowledgeSource[] {
    const result = action.status === "pending" || action.status === "rejected" ? record(action.result.preview) : action.result;
    return Object.entries(KINDS).flatMap(([key, kind]) => {
        const data = record(result[key]);
        const id = Number(data.id ?? action.arguments[`${key}_id`]);
        const user = record(data.user);
        const title = text(data.title) ?? text(data.name) ?? text(data.original_name) ?? text(data.body) ?? text(data.body_md)
            ?? (Object.keys(user).length ? [user.last_name, user.first_name].filter(Boolean).join(" ") || text(user.username) : null);
        if (!title || !Number.isSafeInteger(id) || id <= 0) return [];
        const taskId = kind === "task" ? id : Number(data.task_id) || null;
        const stage = record(data.stage);
        return [{ source_id: `${kind}:${id}`, entity_type: kind, entity_id: id, title, excerpt: null, score: null,
            task_id: taskId, document_slug: text(data.slug),
            card: { key: text(data.key) ?? (kind === "risk" ? `RISK-${id}` : null), status: text(data.stage_name) ?? text(stage.name) ?? text(data.status), priority: text(data.priority), assignee: text(data.assignee), due_date: text(data.due_date), summary: text(data.description_md)?.slice(0, 240) ?? null },
        }];
    });
}

File: backend/src/services/agent_cards.py
"""Представление источников агента: только разрешённые поля текущего каталога."""

from src.knowledge.catalog import ProjectCatalog, ProjectSource, SourceType
from src.schemas.knowledge import KnowledgeSourceCardSchema


def source_card(source: ProjectSource, catalog: ProjectCatalog) -> KnowledgeSourceCardSchema:
    """Собирает карточку из того же PostgreSQL-среза, которым проверены ссылки."""
    data = source.properties
    project = catalog.sources.get(f"project:{source.project_id}")
    key = None
    status = data.get("status")
    assignee = data.get("assignee")
    if source.kind is SourceType.TASK:
        if project and data.get("number") is not None:
            key = f"{project.properties.get('key', '')}-{data['number']}"
        stage = catalog.sources.get(f"stage:{data.get('stage_id')}")
        status = stage.title if stage else None
        user_ids = {
            row["user_id"]
            for row in catalog.rows.get("task_participants", [])
            if row.get("task_id") == source.entity_id and row.get("role") == "EXECUTOR"
        }
        names = [
            " ".join(user.get(part) or "" for part in ("last_name", "first_name")).strip()
            or user.get("username", "")
            for user in catalog.rows.get("users", [])
            if user["id"] in user_ids
        ]
        assignee = ", ".join(names) or assignee
    elif source.kind is SourceType.RISK:
        key = f"RISK-{source.entity_id}"
    elif source.kind is SourceType.PROJECT:
        key = data.get("key")
    return KnowledgeSourceCardSchema(
        key=key,
        status=status,
        priority=data.get("priority"),
        assignee=assignee,
        due_date=str(data["due_date"]) if data.get("due_date") else None,
        summary=(data.get("description_md") or "")[:240] or None,
    )
